_posts/2015-02-20-elasticsearch-and-denormalization.md

```diff
@@ -0,0 +1,223 @@
+analysis, photography, videography etc. Again this project would require loading
+multiple rows of data built from many related Postgres tables, only this time
+searching, sorting and the ability to see many days in a single view were much
+more important features.
+
+We had been using [ElasticSearch][elasticsearch] for a couple of years and never really taken
+advantage of it as a data store (as opposed to using it as a search engine and
+loading ActiveRecord objects found from the primary key ids it returned). The
+advantages were obvious: ElasticSearch indexes are very easy to search and 
+with no ActiveRecord objects to instantiate and no need to hit the database the
+response time for even large numbers of records would be acceptable.
+
+[elasticsearch]: http://www.elasticsearch.org/
+
+All we needed to do was build a search index that contained all the data we
+wanted to display in Sample Queue view layer. However, we realised it would be
+useful if we could swap out individual rows of the Sample Queue with
+ActiveRecord objects or JSON hashes. Or, to put it another way, the view layer
+shouldn't care about the class of the object it was rendering.
+
+### Presenters as Denormalizers
+The [Presenter pattern][presenter-pattern] is a familiar one in Rails: wrap
+objects in a presenter class to keep business logic away from your HAML or ERB.
+Since the process of building up a hash to push into an ElasticSearch index is basically the same
+we decided to use one class to do both jobs. The `SampleDenormalizer` looks like
+this:
+
+[presenter-pattern]: http://nithinbekal.com/posts/rails-presenters/ "Nithin Bekal on Presenters in Rails"
+
+```ruby
+class SampleDenormalizer
+
+  attr_reader :sample
+  def initializer(sample)
+    @sample = sample
+  end
+
+  def to_hash
+    # this list of methods_to_hash can get very long but it encourages you to
+    # keep the hash flat and to be deliberate about what you are adding in there
+    methods_to_hash = %w(id display_name original_image has_basic_attributes?)
+    methods_to_hash += %w(style_id style_id style_name color_name)
+    # etc ...
+    Hash[methods_to_hash.map{|mth| [mth, self.send(mth)]}]
+  end
+
+  def display_name
+    sample.display_name
+  end
+
+  def original_image
+    sample.original_image
+  end
+
+  def color_name
+    sample.color.name
+  end
+
+  def style_name
+    style.name
+  end
+
+  def style_id
+    style.id
+  end
+
+  # One of the powers of this pattern is for attributes like this
+  # This method returns a boolean value that is complex and expensive to
+  # calculate
+  # What's more, it is nearly impossible to query with SQL.
+  # Once you put it in the ElasticSearch index it is cheap to store and return
+  # and simple to query
+  def has_basic_attributes?
+    return false unless style_scalar_attributes
+    required_scalar_values     = REQUIRED_BASIC_SCALAR_ATTRIBUTES.map{|attr| style_scalar_attributes.send(attr) }
```

Comment:
Where is `REQUIRED_BASIC_SCALAR_ATTRIBUTES` coming from?

I think this method might be a confusing example, because to us, "attributes" means something very different than what is created by `attr_accessor`.  I think readers will assume "attributes"  means the later, and wonder if this is some sort of meta-validation we are doing.